File: py/aoc2022/day9.py
"""
Day 9: Rope Bridge
"""


def _moves(lines):
    current_x, current_y = 0, 0
    yield current_x, current_y
    for line in lines:
        direction, count = line.split()
        for _ in range(int(count)):
            match direction:
                case "L":
                    current_x -= 1
                case "R":
                    current_x += 1
                case "U":
                    current_y -= 1
                case "D":
                    current_y += 1
            yield current_x, current_y


def _follow(points):
    tail_x, tail_y = 0, 0
    yield tail_x, tail_y
    for head_x, head_y in points:
        delta_x, delta_y = head_x - tail_x, head_y - tail_y
        if abs(delta_x) <= 1 and abs(delta_y) <= 1:
            continue
        if abs(delta_x) < abs(delta_y):
            tail_x, tail_y = head_x, head_y - delta_y // abs(delta_y)
        elif abs(delta_x) > abs(delta_y):
            tail_x, tail_y = head_x - delta_x // abs(delta_x), head_y
        else:
            tail_x, tail_y = head_x - delta_x // abs(delta_x), head_y - delta_y // abs(
                delta_y
            )
        yield tail_x, tail_y


def part1(lines):
    """
    >>> part1(["R 4", "U 4", "L 3", "D 1", "R 4", "D 1", "L 5", "R 2"])
    13
    """
    return len(set(_follow(_moves(lines))))


def part2(lines):
    """
    >>> part2(["R 5", "U 8", "L 8", "D 3", "R 17", "D 10", "L 25", "U 20"])
    36
    """
    # fmt: off
    # pylint: disable=line-too-long
    return len(set(_follow(_follow(_follow(_follow(_follow(_follow(_follow(_follow(_follow(_moves(lines))))))))))))
    # fmt: on


parts = (part1, part2)

File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2022/Day9.kt
package com.github.ephemient.aoc2022

import kotlin.math.absoluteValue
import kotlin.math.sign

@Day
class Day9(lines: List<String>) {
    private val moves = sequence {
        var x = 0
        var y = 0
        yield(x to y)
        for (line in lines) {
            when (line[0]) {
                'L' -> repeat(line.drop(2).toInt()) { yield(--x to y) }
                'R' -> repeat(line.drop(2).toInt()) { yield(++x to y) }
                'U' -> repeat(line.drop(2).toInt()) { yield(x to --y) }
                'D' -> repeat(line.drop(2).toInt()) { yield(x to ++y) }
            }
        }
    }

    @Day.Part
    fun part1(): Int = moves.follow().toSet().size

    @Day.Part
    fun part2(): Int = moves
        .follow() // 1
        .follow() // 2
        .follow() // 3
        .follow() // 4
        .follow() // 5
        .follow() // 6
        .follow() // 7
        .follow() // 8
        .follow() // 9
        .toSet()
        .size
}

private fun Sequence<IntPair>.follow(): Sequence<IntPair> = sequence {
    var tailX = 0
    var tailY = 0
    yield(tailX to tailY)
    for ((headX, headY) in this@follow) {
        val deltaX = headX - tailX
        val deltaY = headY - tailY
        if (deltaX.absoluteValue <= 1 && deltaY.absoluteValue <= 1) continue
        tailX = if (deltaX.absoluteValue >= deltaY.absoluteValue) headX - deltaX.sign else headX
        tailY = if (deltaX.absoluteValue <= deltaY.absoluteValue) headY - deltaY.sign else headY
        yield(tailX to tailY)
    }
}
